src/mcts/search.cc

```diff
@@ -1218,6 +1224,19 @@ void SearchWorker::GatherMinibatch2() {
     // If there's something to process without touching slow neural net, do it.
     if (minibatch_size > 0 && computation_->GetCacheMisses() == 0) return;
 
+    // If there is backend work to be done, and the backend is idle - exit
+    // immediately.
+    // Only do this fancy work if there are multiple threads as otherwise we
+    // early exit from every batch since there is never another search thread to
+    // be keeping the backend busy. Which would mean that threads=1 has a
+    // massive nps drop.
+    if (multithreaded && minibatch_size > 0 &&
+        computation_->GetCacheMisses() > 0 &&
+        search_->backend_waiting_counter_.load(std::memory_order_relaxed) ==
+            0) {
```

Comment:
We can make one of the limits here configurable via an option, as a way to tune and disable the optimization.